packages/grpc-js-core/src/make-client.ts

```diff
@@ -0,0 +1,178 @@
+export interface ServiceDefinition {
+  [index: string]: MethodDefinition<object, object>;
+}
+
+export interface PackageDefinition {
+  [index: string]: ServiceDefinition;
+}
+
+function getDefaultValues<T>(metadata?: Metadata, options?: T): {
+  metadata: Metadata;
+  options: Partial<T>;
+} {
+  return {
+    metadata: metadata || new Metadata(),
+    options: options || {}
+  };
+}
+
+/**
+ * Map with wrappers for each type of requester function to make it use the old
+ * argument order with optional arguments after the callback.
+ * @access private
+ */
+const deprecated_request_wrap = {
+  unary: (makeUnaryRequest: Function) => {
+    return function makeWrappedUnaryRequest<RequestType, ResponseType>(
+        this: Client, argument: RequestType,
+        callback: UnaryCallback<ResponseType>, metadata?: Metadata,
+        options?: CallOptions) {
+      const opt_args = getDefaultValues(metadata, options);
+      return makeUnaryRequest.call(this, argument, opt_args.metadata,
+                                    opt_args.options, callback);
+    };
+  },
+  client_stream: (makeServerStreamRequest: Function) => {
+    return function makeWrappedClientStreamRequest<RequestType, ResponseType>(
+        this: Client, callback: UnaryCallback<ResponseType>,
+        metadata?: Metadata, options?: CallOptions) {
+      const opt_args = getDefaultValues(metadata, options);
+      return makeServerStreamRequest.call(this, opt_args.metadata,
+                                          opt_args.options, callback);
+    };
+  },
+  server_stream: (f: Function) => f,
+  bidi: (f: Function) => f
+};
+
+/**
+ * Map with short names for each of the requester maker functions. Used in
+ * makeClientConstructor
+ * @private
+ */
+const requester_funcs = {
+  unary: Client.prototype.makeUnaryRequest,
+  server_stream: Client.prototype.makeServerStreamRequest,
+  client_stream: Client.prototype.makeClientStreamRequest,
+  bidi: Client.prototype.makeBidiStreamRequest
+};
+
+export type ServiceClientConstructor = typeof Client & {
+  service: ServiceDefinition;
+};
+
+/**
+ * Creates a constructor for a client with the given methods, as specified in
+ * the methods argument. The resulting class will have an instance method for
+ * each method in the service, which is a partial application of one of the
+ * [Client]{@link grpc.Client} request methods, depending on `requestSerialize`
+ * and `responseSerialize`, with the `method`, `serialize`, and `deserialize`
+ * arguments predefined.
+ * @param methods An object mapping method names to
+ *     method attributes
+ * @param serviceName The fully qualified name of the service
+ * @param classOptions An options object.
+ * @return New client constructor, which is a subclass of
+ *     {@link grpc.Client}, and has the same arguments as that constructor.
+ */
+export function makeClientConstructor(
+    methods: ServiceDefinition, serviceName: string,
+    classOptions: ClientClassOptions): ServiceClientConstructor {
```

Comment:
I don't even know if this makes sense, but is it possible to specify that the `methods` argument to this function is the same object as the `service` property attached to the return value?

Reply:
Hmm... I don't think TS is expressive enough for that.